// @openzeppelin/truffle-upgrades/dist/utils/factories.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUpgradeableBeaconFactory = exports.getBeaconProxyFactory = exports.getProxyAdminFactory = exports.getITransparentUpgradeableProxyFactory = exports.getTransparentUpgradeableProxyFactory = exports.getProxyFactory = void 0;
const truffle_1 = require("./truffle");
const ERC1967Proxy_json_1 = __importDefault(require("@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol/ERC1967Proxy.json"));
const TransparentUpgradeableProxy_json_1 = __importDefault(require("@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol/TransparentUpgradeableProxy.json"));
const ITransparentUpgradeableProxy_json_1 = __importDefault(require("@openzeppelin/upgrades-core/artifacts/@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol/ITransparentUpgradeableProxy.json"));
const ProxyAdmin_json_1 = __importDefault(require("@openzeppelin/upgrades-core/artifacts//@openzeppelin/contracts/proxy/transparent/ProxyAdmin.sol/ProxyAdmin.json"));
const BeaconProxy_json_1 = __importDefault(require("@openzeppelin/upgrades-core/artifacts//@openzeppelin/contracts/proxy/beacon/BeaconProxy.sol/BeaconProxy.json"));
const UpgradeableBeacon_json_1 = __importDefault(require("@openzeppelin/upgrades-core/artifacts//@openzeppelin/contracts/proxy/beacon/UpgradeableBeacon.sol/UpgradeableBeacon.json"));
function makeFactoryGetter(artifacts) {
    return function (template) {
        const contract = (0, truffle_1.TruffleContract)(artifacts);
        contract.setProvider(template?.currentProvider ?? (0, truffle_1.getTruffleProvider)());
        contract.defaults(template?.class_defaults ?? (0, truffle_1.getTruffleDefaults)());
        return contract;
    };
}
exports.getProxyFactory = makeFactoryGetter(ERC1967Proxy_json_1.default);
exports.getTransparentUpgradeableProxyFactory = makeFactoryGetter(TransparentUpgradeableProxy_json_1.default);
exports.getITransparentUpgradeableProxyFactory = makeFactoryGetter(ITransparentUpgradeableProxy_json_1.default);
exports.getProxyAdminFactory = makeFactoryGetter(ProxyAdmin_json_1.default);
exports.getBeaconProxyFactory = makeFactoryGetter(BeaconProxy_json_1.default);
exports.getUpgradeableBeaconFactory = makeFactoryGetter(UpgradeableBeacon_json_1.default);
